runtime/compiler/runtime/SymbolValidationManager.cpp

```diff
@@ -0,0 +1,2006 @@
+
+   TR_YesNoMaybe isInstanceOf = fej9->isInstanceOf(classOne, classTwo, objectTypeIsFixed, castTypeIsFixed);
+
+   return (wasInstanceOf == isInstanceOf);
+   }
+
+bool
+TR::SymbolValidationManager::validateSystemClassByNameRecord(uint16_t systemClassID, J9ROMClass *romClass)
+   {
+   TR::Compilation* comp = TR::comp();
+   J9VMThread *vmThread = comp->j9VMThread();
+   TR_J9VM *fej9 = (TR_J9VM *)TR_J9VMBase::get(vmThread->javaVM->jitConfig, vmThread);
+
+   J9UTF8 * className = J9ROMCLASS_CLASSNAME(romClass);
+   TR_OpaqueClassBlock *systemClassByName = fej9->getSystemClassFromClassName(reinterpret_cast<const char *>(J9UTF8_DATA(className)),
+                                                                              J9UTF8_LENGTH(className));
+
+   int32_t arrayDims = 0;
+   systemClassByName = getBaseComponentClass(systemClassByName, arrayDims);
+
+   return validateSymbol(systemClassID, static_cast<void *>(systemClassByName));
+   }
+
+bool
+TR::SymbolValidationManager::validateClassFromITableIndexCPRecord(uint16_t classID, uint16_t beholderID, uint32_t cpIndex)
+   {
+   TR_ASSERT_FATAL(getSymbolFromID(beholderID) != 0, "beholderID %u should exist\n", beholderID);
+
+   TR::Compilation* comp = TR::comp();
+   J9VMThread *vmThread = comp->j9VMThread();
+   TR_J9VM *fej9 = (TR_J9VM *)TR_J9VMBase::get(vmThread->javaVM->jitConfig, vmThread);
+
+   J9Class *beholder = static_cast<J9Class *>(getSymbolFromID(beholderID));
+   J9ConstantPool *beholderCP = J9_CP_FROM_CLASS(beholder);
+   uintptr_t pITableIndex;
+   TR_OpaqueClassBlock *clazz = TR_ResolvedJ9Method::getInterfaceITableIndexFromCP(fej9, beholderCP, cpIndex, &pITableIndex);
+
+   int32_t arrayDims = 0;
+   clazz = getBaseComponentClass(clazz, arrayDims);
+
+   return validateSymbol(classID, static_cast<void *>(clazz));
+   }
+
+bool
+TR::SymbolValidationManager::validateDeclaringClassFromFieldOrStaticRecord(uint16_t definingClassID, uint16_t beholderID, int32_t cpIndex)
+   {
+   TR_ASSERT_FATAL(getSymbolFromID(beholderID) != 0, "beholderID %u should exist\n", beholderID);
+
+   TR::Compilation* comp = TR::comp();
+   J9VMThread *vmThread = comp->j9VMThread();
+   TR_J9VM *fej9 = (TR_J9VM *)TR_J9VMBase::get(vmThread->javaVM->jitConfig, vmThread);
+
+   J9Class *beholder = static_cast<J9Class *>(getSymbolFromID(beholderID));
+   J9ROMClass *beholderRomClass = beholder->romClass;
+   J9ConstantPool *beholderCP = J9_CP_FROM_CLASS(beholder);
+   J9ROMFieldRef *romCPBase = (J9ROMFieldRef *)((UDATA)beholderRomClass + sizeof(J9ROMClass));
+
+   int32_t classCPIndexOfFieldOrStatic = -1;
+   if (cpIndex != -1)
+      classCPIndexOfFieldOrStatic = ((J9ROMFieldRef *)(&romCPBase[cpIndex]))->classRefCPIndex;
+
+   J9Class *definingClass = NULL;
+   J9Class *cpClass = (J9Class*)TR_ResolvedJ9Method::getClassFromCP(fej9, beholderCP, comp, classCPIndexOfFieldOrStatic);
+
+   if (cpClass)
+      {
+      TR::VMAccessCriticalSection getDeclaringClassFromFieldOrStatic(fej9);
+
```

Comment:
Just picking a random example here. It seems like this code might duplicate some logic present in the JIT somewhere....what's the model for keeping the conditions in this code maintained alongside the equivalent logic elsewhere? e.g. if someone finds a bug in code that looks up the declaring class, do we now need to fix that in more than one place? I don't have any specific examples, but much of this code, for example, "feels" like code in the front end...

Reply:
> what's the model for keeping the conditions in this code maintained alongside the equivalent logic elsewhere?

I want to remove duplicate versions of code like this and have one canonical version. Time constraints resulted in duplicated versions, but when I wrote this out, I was cognizant of the fact that in the future I would have to consolidate all of this in order to prevent exactly the situation you're worried about.